rustc_mir_build: Enforce `rustc::potential_query_instability` lint

Enselic opened the pull request:
Stop allowing `rustc::potential_query_instability` on all of `rustc_mir_build` and instead allow it on a case-by-case basis if it is safe to do so. In this crate there was only one instance of the lint, and it was safe to allow.

Part of https://github.com/rust-lang/rust/issues/84447 which is E-help-wanted.

cjgillot commented on compiler/rustc_mir_build/src/thir/pattern/check_match.rs:
```diff
@@ -952,6 +952,9 @@ fn report_non_exhaustive_match<'p, 'tcx>(
+        // The items are only used to add notes to an err, so the order does not
+        // affect query stability.
+        #[allow(rustc::potential_query_instability)]
         for ty in non_exhaustive_tys {
```
The order of diagnostics is observable for the user, so having it platform independent would be better.

michaelwoerister replied:
Using `FxIndexSet` instead of `FxHashSet` for `non_exhaustive_tys` should achieve that goal.

Enselic committed:
rustc_mir_build: Make non-exhaustive non-empty match diagnotics deterministic

Enselic committed:
rustc_mir_build: Enforce `rustc::potential_query_instability` lint

Stop allowing `rustc::potential_query_instability` on all of
`rustc_mir_build` and instead allow it on a case-by-case basis if it is
safe to do so. In this crate there was no instance of the lint
remaining.

michaelwoerister commented on compiler/rustc_mir_build/src/thir/pattern/check_match.rs:
```diff
@@ -948,10 +948,15 @@ fn report_non_exhaustive_match<'p, 'tcx>(
         collect_non_exhaustive_tys(cx, &witnesses[0], &mut non_exhaustive_tys);
 
+        // Make order of diagnostics deterministic.
+        non_exhaustive_tys.sort();
```
FxIndexSet will maintain its insertion order as the iteration order. Since it is filled in a deterministic way, we don't need to do any sorting (and also don't need to `#[allow]` the lint below).

Enselic replied:
Oops, good catch, thanks. Done.

michaelwoerister commented:
Thanks, @Enselic!
@bors r+ rollup

bors commented:
:pushpin: Commit 5644a53426a36159d17c8b7135a41a0be059af45 has been approved by `michaelwoerister`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=michaelwoerister 5644a53426a36159d17c8b7135a41a0be059af45 -->
<!-- homu: {"type":"Approved","sha":"5644a53426a36159d17c8b7135a41a0be059af45","approver":"michaelwoerister","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
